src/luxos/asyncops.py

```diff
@@ -0,0 +1,208 @@
+            if asjson:
+                return json.loads(res)
+            else:
+                return res
+        except (Exception, asyncio.TimeoutError) as e:
+            last_exception = e
+        if retry_delay:
+            await asyncio.sleep(retry_delay)
+        count += 1
+    if last_exception is not None:
+        raise last_exception
+
+
+def validate_message(
+    host: str,
+    port: int,
+    res: dict[str, Any],
+    extrakey: str | None = None,
+    minfields: None | int = None,
+    maxfields: None | int = None,
+) -> dict[str, Any]:
+    # all miner message comes with a STATUS
+    for key in ["STATUS", "id", *([extrakey] if extrakey else [])]:
+        if key in res:
+            continue
+        raise exceptions.MinerCommandMalformedMessageError(host, port, f"missing {key} from logon message", res)
+
+    if not extrakey or not (minfields or maxfields):
+        return res
+
+    n = len(res[extrakey])
+    msg = None
+    if minfields and (n < minfields):
+        msg = f"found {n} fields for {extrakey} invalid: " f"{n} <= {minfields}"
+    elif maxfields and (n > maxfields):
+        msg = f"found {n} fields for {extrakey} invalid: " f"{n} >= {maxfields}"
+    if msg is None:
+        return res[extrakey]
+    raise exceptions.MinerCommandMalformedMessageError(host, port, msg, res)
+
+
+@wrapped
+async def logon(host: str, port: int, timeout: float | None = 3) -> str:
+    timeout = TIMEOUT if timeout is None else timeout
+    res = await roundtrip(host, port, {"command": "logon"}, timeout)
+
+    # when we first logon, we'll receive a token (session_id)
+    #   [STATUS][SessionID]
+    # on subsequent logon, we receive a
+    #   [STATUS][Msg] == "Another session is active" ([STATUS][Code] 402)
+    if "SESSION" not in res and res.get("STATUS", [{}])[0].get("Code") == 402:
+        raise exceptions.MinerCommandSessionAlreadyActive(host, port, "connection active", res)
+    sessions = validate_message(host, port, res, "SESSION", 1, 1)
+
+    session = sessions[0]  # type: ignore
+
+    if "SessionID" not in session:
+        raise exceptions.MinerCommandSessionAlreadyActive(host, port, "no SessionID in data", res)
+    return str(session["SessionID"])
+
+
+@wrapped
+async def logoff(
+    host: str, port: int, sid: str, timeout: float | None = 3
+) -> dict[str, Any]:
+    timeout = TIMEOUT if timeout is None else timeout
+    return await roundtrip(host, port, {"command": "logoff", "parameter": sid}, timeout)
+
+
+@wrapped
+async def execute_command(
+    host: str,
+    port: int,
+    timeout_sec: float | None,
+    cmd: str,
+    parameters: list[str] | None = None,
+    verbose: bool = False,
+    asjson: bool | None = True,
+    add_address: bool = False
+) -> tuple[tuple[str, int], dict[str, Any]] | dict[str, Any]:
+    timeout = TIMEOUT if timeout_sec is None else timeout_sec
+    parameters = parameters or []
+
+    sid = None
+    if api.logon_required(cmd):
```

Comment:
There were some cases on other tools we developed that triggering a new logon could cause an issue if there was already an active session for the miner. We have another command https://docs.luxor.tech/firmware/api/luxminer/session that retrieves the session id that can be used instead of trying a new logon. Just FYI.

Reply:
My mental model of a session acting as a lock like this, plese let me know if that's wrong:
```
   client                               miner
  (1) start session   -------> (A) create "session"
  (2) set val 99    ----------> |    start set val 99
  (3) set val 100  ----------X | 
                                            |    commit 99
  (4) close session --------> (B) release "session"

```
But both ends can die. If miner dies  (B) won't release the session but that's not a big issue, because there won't be a session (unless is not persisted miner side?), 
But if the client dies, then (4) will not happen and furter command will have to wait for the ttl to expire the session.